prov/off_s - SHARP provider scaffolding

grom72 opened the pull request:
Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

Requires #2 

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/libfabric/1)
<!-- Reviewable:end -->

grom72 committed:
prov/off_s - SHARP provider scaffolding

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/coll: Prepare coll_cq to be used in by collective providers

coll_cq implementation can be reused by other collective providers.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm - collective offload capabilities reported if offload provider is available

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Collective offload fabric setup integrated into rxm fabric initialization

It is rxm provider responsability to initialize collective offload provider's fabric.
Otherwise collective offload functionality will not be available

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
core/fabric: Collective offload provider set via FI_OFFLOAD_COLL_PROVIDER

FI_OFFLOAD_PROVIDER environment variable shall be set to offload provider name
to instruct libcabric to setup and use particular provider.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Create eq for collective offload provider

Peer provider must create peer_eq for offload provider, to allow offload provider
reporting events to peer provider.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/1)* status: 0 of 30 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

grom72 committed:
prov/rxm: FI_PEER flag force fi_join() with util_coll_ep

Offload provider may execute collective operations via util_coll provider.
It must call fi_join() operation to get struct mc required for collective operations.
It can only call fi_join() on it's peer provider (e.g. rxm). FI_PEER flag is used
to inform peer provider to coll fi_join() operation for util_coll_ep

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

j-xiong committed:
prov/rxm: Always use rendezvous protocol for ZE device memory send

Bounce buffer copy overhead is high for ZE device memory. The rendezvous
protocol takes advantage of GPU RDMA and performs better even for small
messages.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

Commit:
fabric: Remove FI_PEER_DOMAIN

The FI_PEER flag is sufficient for allocating peer objects.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

Commit:
prov/net: Remove unused ep::hit_cnt field

This is carry over from the tcp provider but unused.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

grom72 committed:
prov/rxm: Set offload_coll_mask based on actual configuration

offload_coll_mask value is calculated based on the actual offload capabilities
confirmed by fi_query_collective().

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau reviewed (changes requested):
> Reviewed 2 of 8 files at r1, 6 of 24 files at r3, 1 of 14 files at r5, 2 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/1)* status: 11 of 34 files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`include/ofi_sharp.h` line 43 at r6](https://reviewable.io/reviews/pmem/libfabric/1#-NIXC8KR4eUhJ9-Cr9LU:-NIXC8KR4eUhJ9-Cr9LV:b2y7yo) ([raw file](https://github.com/pmem/libfabric/blob/a2f7709f8833f51ae0a177407b07b00361d3c82a/include/ofi_sharp.h#L43)):*
> ```objc
> 
> #include <ofi_atom.h>
> #include <ofi_proto.h>
> ```

It seems they all are not needed here ...



_Code quote:_
```objc
#include "config.h"

#include <stdint.h>
#include <stddef.h>
#include <sys/un.h>

#include <ofi_atom.h>
#include <ofi_proto.h>
#include <ofi_mem.h>
#include <ofi_rbuf.h>
#include <ofi_tree.h>
#include <ofi_hmem.h>

#include <rdma/providers/fi_prov.h>

#include "ofi_coll.h"
```
___
*[`include/rdma/fi_domain.h` line 252 at r6](https://reviewable.io/reviews/pmem/libfabric/1#-NIXCJq2DFW9hjf5s_sa:-NIXCJq2DFW9hjf5s_sb:bi6hj6) ([raw file](https://github.com/pmem/libfabric/blob/a2f7709f8833f51ae0a177407b07b00361d3c82a/include/rdma/fi_domain.h#L252)):*
> ```objc
> 	FI_SCATTER,
> 	FI_GATHER,
> 	FI_COLLECTIVE_OP_MAX, //must be always the last value of this enum
> ```

`MAX /* must be always the last value of this enum */`



_Code quote:_
```objc
MAX, //must be always the last value of this enum
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> Reviewed 6 of 36 files at r7, 1 of 1 files at r8, 15 of 30 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/1)* status: 22 of 37 files reviewed, 6 unresolved discussions (waiting on @grom72)

___
*[`prov/rxm/src/rxm_ep.c` line 394 at r9](https://reviewable.io/reviews/pmem/libfabric/1#-NIXUwP58xH8UMUk4Xr_:-NIXUwP58xH8UMUk4Xra:b-3ekxo8) ([raw file](https://github.com/pmem/libfabric/blob/3a0464954a6728fee9b7a7dcae161dd497e5f33c/prov/rxm/src/rxm_ep.c#L394)):*
> ```C
> 
> 	//FI_PEER flag is used to force util_coll context 
> 	//where fi_join() is called from offload provider
> ```

1) C++ style comment
2) space at the end



_Code quote:_
```C
	//FI_PEER flag is used to force util_coll context
	//where fi_join() is called from offload provider
```
___
*[`prov/rxm/src/rxm_ep.c` line 402 at r9](https://reviewable.io/reviews/pmem/libfabric/1#-NIXV0nVDSS6X-tEYzOH:-NIXV0nVDSS6X-tEYzOI:b3kqx) ([raw file](https://github.com/pmem/libfabric/blob/3a0464954a6728fee9b7a7dcae161dd497e5f33c/prov/rxm/src/rxm_ep.c#L402)):*
> ```C
> 		if (ret)
> 			goto err_util_coll;
> 		// It is collective offload provider responsibility to store util_coll provider mc
> ```

.



_Code quote:_
```C
// It is collective offload provider responsibility to store util_coll provider mc
```
___
*[`prov/rxm/src/rxm_msg.c` line 818 at r9](https://reviewable.io/reviews/pmem/libfabric/1#-NIXV80kDT80rBlOP7jL:-NIXV80kDT80rBlOP7jM:bdf2pec) ([raw file](https://github.com/pmem/libfabric/blob/3a0464954a6728fee9b7a7dcae161dd497e5f33c/prov/rxm/src/rxm_msg.c#L818)):*
> ```C
> 				   rxm_ep_sar_calc_segs_cnt(rxm_ep, data_len));
> 	} else {
> rndv_send:
> ```

Empty line before label, not after



_Code quote:_
```C
	} else {
rndv_send:

		ret = rxm_alloc_rndv_buf(rxm_ep, rxm_conn, context,
```
___
*[`prov/sharp/configure.m4` line 1 at r9](https://reviewable.io/reviews/pmem/libfabric/1#-NIXVGet1NRF6U3uPxmV:-NIXVGet1NRF6U3uPxmW:b3kre) ([raw file](https://github.com/pmem/libfabric/blob/3a0464954a6728fee9b7a7dcae161dd497e5f33c/prov/sharp/configure.m4#L1)):*
> ```M4Sugar
> dnl Configury specific to the libfabric sharp provider
> ```

?



_Code quote:_
```M4Sugar
Configury
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/1)* status: 22 of 37 files reviewed, 7 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/src/sharp_domain.c` line 124 at r10](https://reviewable.io/reviews/pmem/libfabric/1#-NIawrRO2NUidj7cyCWo:-NIawrRPBlgi3Fiu_Dnv:bvjqt2r) ([raw file](https://github.com/pmem/libfabric/blob/d288bc663b1ea8279eec5281112591bf00595a0b/prov/sharp/src/sharp_domain.c#L124)):*
> ```C
> 	/// mapped to int sharp_coll_finalize(struct sharp_coll_context *context);
> 	ret = ofi_domain_close(&domain->util_domain);
> 	if (ret)
> ```

Memory leak - domain is not freed.



_Code quote:_
```C
	if (ret)
		return ret;
```


<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/libfabric/1)* status: 22 of 37 files reviewed, 6 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/src/sharp_domain.c` line 124 at r10](https://reviewable.io/reviews/pmem/libfabric/1#-NIawrRO2NUidj7cyCWo:-NIbG40PA-_8R_GMTADi:b9o3oa) ([raw file](https://github.com/pmem/libfabric/blob/d288bc663b1ea8279eec5281112591bf00595a0b/prov/sharp/src/sharp_domain.c#L124)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Memory leak - domain is not freed.
</blockquote></details>

Maybe not a memory leak - if .close checks for return code and retries - I hope it does so ;-)


<!-- Sent from Reviewable.io -->

grom72 committed:
prov/sharp: Point of fi_barier() implementation added

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 commented:
see #8
